### ISSUE TITLE
Digispark 2-pin hookup source voltage and I/O voltage + timing issue

### ISSUE THREAD

**Issue by nb-programmer** [dparson55/NRFLite#61]
The ATTiny85 uses 5V on the Digispark, and the nRF uses 3.3V for power. So, the I/O on the ATTiny would be 5V as well, which is perfectly suited to interface the nRF module. But the [2-pin hookup](https://github.com/dparson55/NRFLite#2-pin-hookup-guide) schematic doesn't seem to work on the default timings on the Digispark board.
Should the ATTiny also run on a 3.3V source for the circuit to work? If it is, how would I adjust the timing RC circuit to work on 5V I/O?

Edit: Found out that #55 is related to this. But I can't run the Digispark on 3.3V, as I need USB functionality

**Comment by dparson55** [comment]
Hi there, using 5 or 3.3V for the 2-pin hookup does not matter, the pin toggling that occurs is so slow that the slightly larger voltage transition doesn't make a difference.  I've tested ATmegas and ATtinys running from 1-16MHz at 3.3 and 5V and all work fine, so unfortunately it's just something to do with the Digispark.

I never did get a Digispark so won't be able to debug it, but here is one idea to try:  replace the call to `noInterrupts();` on [line 553](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L553) with `cli();` and replace the call to `interrupts();` on [line 592](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L592) with `sei();`.

The idea with this change is maybe the Digispark core is overriding the normal behavior of `noInterrupts` due to some USB interrupt requirements.  I couldn't find the code to determine if their core is doing this so it's just a theory.  If there is still an interrupt that is allowed after `noInterrupts` is called, it would break the delicate timing required for the SPI bit-banging that NRFLite is performing.  But this change might break the USB functionality so even if this fixes the radio communication, it still might not be a usable solution.

If this doesn't help another idea to try is slowing down the interactions NRFLite does with the radio by doing this:  replace the 1K resistor with a 2K resistor (or two 1K's in series) and replace the 0.1uF capacitor with a 0.2uF capacitor (or two 0.1uF in parallel).  Then change [line 558](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L558) and [line 565](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L565) to be `delayMicroseconds(1000);`.  Normally the delay is 500 microseconds so all we are doing is doubling everything to slow things down.

If none of this works but you have a 4 channel oscilloscope, let me know and we can continue the troubleshooting.

**Comment by nb-programmer** [comment]
I have an update, sorry for the long delay
I changed the lines as you said, but to no avail. The USI version works, but the only problem is there are no more pins remaining for USB anymore, as the P5 pin is for RESET only. (I could re-flash the ATTiny fuse bits to disable the RESET function, but that is risky and I don't have the correct programmer. So if all else fails, I will have to try that)
I don't have an analog oscilloscope, but I do have a Logic Analyzer, and managed to capture the interactions.
(Attached the capture with some labels, open it with [Logic 2](https://www.saleae.com/downloads/))
What seems strange is the SCK and CSN signals do not separate at all, and are exactly the same, so there is definitely some timing issue. I didn't try swapping the RC components yet, but I'll update when I do.

[Digispark NRFLite.zip](https://github.com/dparson55/NRFLite/files/6507371/Digispark.NRFLite.zip)

**Comment by dparson55** [comment]
Having that capture is great, I'm still not sure what could be happening but like you mentioned, the CSN and SCK being coupled is a problem.

When NRFLite sends a command to the radio, it brings the 2-pin SCK pin LOW and waits 500 microseconds for the capacitor that's connected to the radio's CE & CSN pins to discharge and become LOW.  I see this [500 microsecond delay](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L558) in the capture.

![image](https://user-images.githubusercontent.com/19499747/118898855-55f1ac80-b8c2-11eb-9d55-ae074fc9ac88.png)

However the CE & CSN should be LOW the entire time SCK is pulsed:  the purpose of the 1K resistor and 0.1uF capacitor is to keep CE & CSN low while we clock data in and out of the radio using SCK pulses.  It seems like this resistor and capacitor are not even present.

What should be happening is the SCK pin is [pulsed HIGH and immediately LOW](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L626) and due to the 1K resistor and 0.1uF capacitor, the CE & CSN pins remain LOW.  The capacitor will charge slightly but the idea is not to let it get past the voltage threshold that represents a HIGH to the radio.  The moment CSN goes HIGH on the radio, it will stop listening to the SPI bus, so the communication with the radio breaks down.

The timing on the SCK pin looks good:  it is HIGH for about 1 microsecond and LOW for about 4 microseconds.  This roughly 20% duty cycle is important in order to prevent the capacitor from charging too high during the SCK clock pulse.  Ideally the voltage on the capacitor would only reach 20% of the IO voltage, so if it's 3.3V it should be no higher than 0.7V and considered LOW by the radio.

![image](https://user-images.githubusercontent.com/19499747/118909706-42e9d700-b8d8-11eb-9e58-d5ce2a21a656.png)

The MOSI pin looks good.  MOSI is being [set prior to the SCK pulse](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L624) and [cleared after the pulse](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L629), and this is how it should look.  I don't think we can read too much into what the MISO pin is doing though.  It is independent of MOSI and that's a good sign (MISO does not go high immediately when MOSI goes high), but since we're failing to clock any data into the radio, we cannot expect to get any data out of it.

I did notice that when SCK goes low, it takes maybe 0.05 nanoseconds before CE & CSN go low.  Maybe this is nothing but I'm not sure.  When SCK goes high, CE & CSN go high at the same time in the capture, and again that should not be happening.

![image](https://user-images.githubusercontent.com/19499747/118913394-3799aa00-b8de-11eb-896e-ba76ea6ca013.png)

Maybe with this additional information you can do some more troubleshooting.  Make sure your CE/CSN probe is connected to the CE & CSN pins of the radio, in between the 1K resistor and 0.1uF capacitor.  Based on the capture it seems like the resistor and capacitor are not doing their job at all, so I'd focus the effort there.

**Comment by dparson55** [comment]
Closing due to lack of responses.  Maybe we can reopen when you have more time.

**Comment by nb-programmer** [comment]
Hi again! Yeah I got caught up with Uni, and also had to wait for replacement caps. Sorry for keeping you waiting.
You were correct on the capacitor being the culprit. The cap I had was garbage and acted as if it was not present (maybe burned out somehow, or was broken from the beginning). So that explains why the SCK and CS was not separating. After replacing the cap, I got some beautiful captures! (Still not working though, but progress!)
![image](https://user-images.githubusercontent.com/46126206/120295331-aca8ae80-c2e4-11eb-8613-525b8d8f80db.png)

(the spurious CSN rise/falls at some places are probably from my Logic analyzer freaking out over the analog voltage, so I doubt that is the issue)

I can now clearly read the individual bits and trace the code. Seems all transmissions are absolutely correct now
Here is a capture with labels: 
[Digispark_NrfLite_Fixed_Cap.zip](https://github.com/dparson55/NRFLite/files/6574858/Digispark_NrfLite_Fixed_Cap.zip)

But after all this, the link was still not establishing. Then I did some more digging and found something suspicious
I wondered if the SPI data was not read correctly by the Digispark. So I modified the NRFLite code to "print" the read data to output pins:
```c
//inside NRFLite::waitForTxToComplete()
...

while (txAttemptCount++ < MAX_TX_ATTEMPT_COUNT)
{
	fifoReg = readRegister(FIFO_STATUS);
	txBufferIsEmpty = fifoReg & _BV(TX_EMPTY);

	//Debug output
	pinMode(PB0, OUTPUT);
	pinMode(PB1, OUTPUT);
	int t = fifoReg;
	for (int i = 0; i < 8; i++) {
		if (t & 0x80) digitalWrite(PB0, HIGH);
		else digitalWrite(PB0, LOW);
		digitalWrite(PB1, HIGH);
		digitalWrite(PB1, LOW);
		t = t << 1;
	}
	digitalWrite(PB0, LOW);


	if (txBufferIsEmpty)
	{
		result = 1; // Indicate success.
		break;
	}
	
	...
```

This will clock out the register contents, which are correctly received, as seen from the above capture.
And look what I got back!
![Incorrect read](https://user-images.githubusercontent.com/46126206/120295690-fee9cf80-c2e4-11eb-83c3-d03c99fc50da.PNG)

[Digispark_NrfLite_Fixed_Cap_Debug.zip](https://github.com/dparson55/NRFLite/files/6574944/Digispark_NrfLite_Fixed_Cap_Debug.zip)
(Just skip to the first annotation)

Bingo! So the reason now it doesn't work is that we're not even getting any input, and the nRF remains uninitialized (according to the code).
Some possible reasons I think this is happening:
- The 10K resistor is too high, and not providing enough voltage to the input pin
- The nRF running at 3.3V and ATTiny at 5V makes it not recognize the lower input voltage
- Maybe both of the above conditions are causing this

I'll try to reduce the 10K to something lower like 4K7 and see if it works.

For reference, here is my connection diagram:
![Diagram](https://user-images.githubusercontent.com/46126206/120299766-f1364900-c2e8-11eb-8baf-98c12434d6a0.png)

Also, here is the main .ino code:
```c
#include <NRFLite.h>

const static uint8_t PIN_RADIO_SCK = PB0;
const static uint8_t PIN_RADIO_MOMI = PB1;

const static uint8_t RADIO_ID = 1;
const static uint8_t DESTINATION_RADIO_ID = 0;

NRFLite _radio;

void setup() {
	if (!_radio.initTwoPin(RADIO_ID, PIN_RADIO_MOMI, PIN_RADIO_SCK, NRFLite::BITRATE250KBPS)) {
		//Cannot communicate with radio.
		while (1) {
			//This delay is kept so the compiler doesn't optimize the loop out
			delay(1000);
			//I had also placed some LED blinking stuff here to confirm it stops here
		}
	}
}

void loop() {
	//Nothing here yet
}
```

Still, nothing can be confirmed till we get oscilloscope readings.

**Comment by dparson55** [comment]
Glad to hear it was just a bad capacitor.  Your connection diagram looks good as well.

I think you are on to something with the MOSI and MISO pins.  After a lot of testing I confirmed your thoughts about the resistor and also have a code change to help with the troubleshooting.

Below is an oscilloscope capture along with the logic analyzer data you provided just for reference.  SCK (labeled CLK) is driven LOW and a 500us wait is performed to allow the capacitor on CS to discharge.  After that point the clocking of data begins.
![image](https://user-images.githubusercontent.com/19499747/120404585-8c381e80-c2fb-11eb-8fcd-da19454c7eba.png)

Zoomed in, this capture shows a 16MHz microcontroller @ 5V, radio @ 3.3V, FIFO_STATUS = 0001 0001, and 10K resistor between MOSI and MISO.  Under these conditions there is not much time for the radio's MISO to force the MOSI pin HIGH.  It is working for me but only if I power the radio with 3.30V:  if it's only 3.29V or below, the 2-pin operation is inconsistent and completely fails if the voltage is below 3.27V.
![image](https://user-images.githubusercontent.com/19499747/120413865-e8a43980-c30d-11eb-8083-ee6155508e05.png)

Here is the same setup using a 5K resistor.  There is still good separation between the microcontroller and radio output pins (red arrows).  The MOSI pin at the point the radio is attempting to output a 1 has a quicker rise (green arrow) compared to using a 10K resistor.  I also had success using this resistor while powering the radio down to 2.75V, so a big improvement over the 3.30V needed when using the 10K resistor.
![image](https://user-images.githubusercontent.com/19499747/120416372-e5ab4800-c311-11eb-888c-9ea727528598.png)

Here is the 2K resistor.  The radio's output pin (MISO) is being distorted quite a bit more, especially compared to using a 10K resistor.  I'm not sure how much the radio's pin is being stressed, but it is seeing 0.8V when it's trying to be 0V, so 2K might be too low.
![image](https://user-images.githubusercontent.com/19499747/120414587-0f16a480-c30f-11eb-9d90-e16cbdfa8e2b.png)

Back to using a 10K resistor, the current twoPinTransfer method results in 1.6us for the MOSI pin to ramp up and be HIGH (the BX-AX measurement).
![image](https://user-images.githubusercontent.com/19499747/120420643-979a4280-c319-11eb-92d6-cb97ff93296e.png)

If you replace the current [twoPinTransfer](https://github.com/dparson55/NRFLite/blob/master/src/NRFLite.cpp#L612) method with the new one below, the ramp up and HIGH time will be extended to 2.2us when using a 10K resistor.  I successfully powered the radio down to 2.8V with this code change.
![image](https://user-images.githubusercontent.com/19499747/120420733-c3b5c380-c319-11eb-83cb-4df912c1e5df.png)
```
uint8_t NRFLite::twoPinTransfer(uint8_t data)
{
    uint8_t byteFromRadio = 0;
    uint8_t currentBitIndex = 8;
    
    do
    {
        byteFromRadio <<= 1; // Shift the byte we are building to the left.
        
        // If the microcontroller is running at 16MHz and using 5V, the 3.3V radio's MISO pin may need
        // additional time to bring the 2-pin circuit's MOMI pin to a voltage the microcontroller will consider HIGH.
        // Using these nops extend the ramp up + HIGH time from 1.6us up to 2.2us.
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t");
        __asm__ __volatile__("nop\n\t"); 

        if (*_momi_PIN & _momi_MASK) { byteFromRadio++; } // Read bit from radio on MOMI pin.  If HIGH, set bit position 0 of our byte to 1.
        *_momi_DDR |= _momi_MASK;                         // Change MOMI to be an OUTPUT pin.
        
        if (data & 0x80) { *_momi_PORT |=  _momi_MASK; }  // Set MOMI HIGH if bit position 7 of the byte we are sending is 1.


        *_sck_PORT |= _sck_MASK;  // Set SCK HIGH to transfer the bit to the radio.  CSN will remain LOW while the capacitor begins charging.
        *_sck_PORT &= ~_sck_MASK; // Set SCK LOW.  CSN will have remained LOW due to the capacitor.
        
        *_momi_PORT &= ~_momi_MASK; // Set MOMI LOW.
        *_momi_DDR &= ~_momi_MASK;  // Change MOMI back to an INPUT.  Since we previously ensured it was LOW, its PULLUP resistor will never 
                                    // be enabled which would prevent MOMI from fully reaching a LOW state.

        data <<= 1; // Shift the byte we are sending to the left.
    }
    while (--currentBitIndex);
    
    return byteFromRadio;
}
```

For microcontrollers running at slower speeds or if the microcontroller and radio are powered by the same voltage, things are good with the 10K resistor and current twoPinTransfer logic.  But for the situation of a Digispark (16MHz, 5V microcontroller, 3.3V radio), I think the 5K resistor or inserting the additional delay is warranted, especially considering how borderline the radio's voltage tested on my end.  Let me know what you find and hopefully just changing to the 5K resistor will be the solution.

**Comment by nb-programmer** [comment]
I have tried your suggested changes using 8 nop's, but it didn't work for me
I then tried increasing the nop's, and then tried delayMicroseconds(from 100 to 500), but still no effect

I then tried to use a 5K MIMO resistor (2 10Ks in parallel), but still the same response. At this point, I'm clueless as to why nothing seems to work.

I took measurements to the supply voltages to make sure, and I got:
- 5V on the USB port connector
- 4.74V on the "5V" pin (maybe there is a reverse-protection diode in place dropping the voltage). This appears even without the nRF connected
- 3.26V on the nRF VCC - this voltage is, as you said, too low to work reliably. I can't do much to adjust this higher, as this is a fixed voltage on my regulator with individual jumpers (there is a 3.3v, 2.5v and 1.8v selection below 5v). Maybe I can use a more precise regulator or LDO to get 3.3v or 3.6v to see if that helps - this may be a reason for it

I also noticed the Digispark runs slightly higher than 16MHz - at 16.5MHz due to the USB timings. Not sure if that is important

![image](https://user-images.githubusercontent.com/46126206/120643524-e66be780-c493-11eb-9abe-f90915ce15a9.png)

I'll try something else with different voltages and resistors and see how it goes

**Comment by dparson55** [comment]
Having 3.26V on the nRF is ok since you are now using the 5K resistor.  In my testing a voltage as low as 2.75V worked when using this lower resistance.  It was only when using a 10K resistor that the communication would fail when the nRF voltage was below 3.30V and the microcontroller was at 5V.  Because of this, I will be updating the 2-pin hookup guide to suggest a 4-6K resistor.

I think the 16.5MHz speed will be ok.  I don't have a way to test it directly but using the 5K resistor provides so much wiggle room for the MIMO pin reading that I don't think this could be a problem.

Can you try a different pin for MOMI?  I noticed you are using PB1 in a previous comment and it looks like this is connected to an LED according to the Digispark schematic below.  Pins PB0, PB2, and PB5 have no connected components so they would be ideal for use with the 2-pin hookup.  If you need to use PB0 and PB1, I'd change SCK to be PB1 and MOMI to be PB0.
![image](https://user-images.githubusercontent.com/19499747/120704716-cfb19a80-c46b-11eb-90c5-27f92e32c06a.png)

You can also try validating the voltage you are powering the nRF with is high enough for the microcontroller to see a 1 when performing a `digitalRead`.  Make the pin you are using for MIMO an INPUT, then connect the 3.26V regulated voltage to a 5K resistor, and the other side of the 5K resistor to your microcontroller pin.  If `digitalRead` returns a 0, it would mean the voltage is not high enough.

You can also temporarily power the nRF with 5V rather than 3.3V.  The maximum supply voltage is listed as 3.6V, but many people, including myself, have accidently given it 5V and it worked ok.  I would not do this permanently but for a quick test, this could help determine if the problem is the MISO pin of the nRF not reaching a high state for the microcontroller.

**Comment by nb-programmer** [comment]
Some good news!
I figured I check if at least there are any hidden factors, and I'm not just tackling a problem that doesn't exist, I tried a different approach.

First I tried to see if the nRF's VCC is high enough to register as a HIGH.
Simple enough, just put pin PB2 as input, and in code just set the state of PB1 (built-in LED). I just poked the pin with a jumper, and the LED reacted. So that works just fine.

Then I tried poking it with a 5K resistor. Didn't expect anything different, and sure enough, the LED pin reacted.
I then desoldered the nRF, and poked the P1 pin as input. And yup, no input, as you said. I guess that LED just drains most of the 10K's (and 5K's) output away. Should've been my first guess

Later on, instead of a 2-pin MOMI approach, I attempted to split the MOMI to separate pins. This way, we will know if the nRF is actually responding and the Digispark reads it at least.

So I modified the code to accept pin P2 to be MISO, and P1 (originally MIMO) to be MOSI. After doing the connections, I modified the code accordingly:
```c++
//inside NRFLite.h
class NRFLite {
...
private:
...
//Added pin mask for MI pin
volatile uint8_t *_mi_PIN;
uint8_t _mi_MASK;

//inside NRFLite.c
//NRFLite::initTwoPin

//Just hard-coding for now
_mi_PIN = portInputRegister(digitalPinToPort(PB2));
_mi_MASK = digitalPinToBitMask(PB2);

//NRFLite::twoPinTransfer
...

if (*_mi_PIN & _mi_MASK) { byteFromRadio++; }

```

And not to my absolute surprise, it started working! Woot!

Also, a thing about the different pins:
You're right about the LED on PB1, which I had guessed would be the problem, that is draining the nRF's output. I did it that way because the nRF SCK and MOSI pins just lines up with the two GPIOs, so I can mount it just on top of the Digispark, like so:
![layout](https://user-images.githubusercontent.com/46126206/120902225-ac017680-c65c-11eb-8197-1a493f6365ce.png)

So I'm thinking of doing one of these things:
- Desolder the LED off the board so I can just use it with two pins, as the LED is barely visible under the nRF anyway
- Try to use USI, but it has it's own caveats:
  - There are actually no remaining free pins left for CE/CSN after the USI port is used. Sure, Pin 5 is there, but as I said, on some Digispark boards (the clone ones), the pin is disabled by default, and is actually the RESET pin. So a complete reflash of the fuses to disable the RESET (Like [here](https://thetoivonen.blogspot.com/2015/12/fixing-pin-p5-or-6-on-digispark-clones.html))
  - Obviously, Pin 3 and 4 are used for USB, so if USB is needed, these pins are unusable

So that leaves me with 3 choices: I can just stick with the "3 pin" version, desolder the LED or use USI by flashing the firmware.
I did do the USI mode first using one of the USB pins, so I know it works. Just need to get one spare pin for /CS.

Anyways, thanks so much for all the help! The tracing analysis actually helped me with debugging! I think there is not much to solve here, maybe to add a note in the README about this
